contracts/TimelockVaultVotes.sol

```diff
@@ -0,0 +1,207 @@
+        Account memory account = accounts[user];
+        uint256 t = block.timestamp;
+
+        // if the time is before the cliff - there's nothing vested yet
+        if (t < account.cliff_timestamp) return 0;
+
+        // if after the cliff
+        uint256 vested;
+        if (account.vesting_ends_timestamp < t) {
+            // if everything is vested
+            vested = uint256(account.amount_under_vesting_div1e12) * 1e12;
+        } else {
+            // otherwise the amount is proportional to the amount after the cliff before end of vesting
+            uint256 x = uint256(account.amount_under_vesting_div1e12) * 1e12;
+            uint256 y = (t - account.cliff_timestamp);
+            uint256 z = (uint256(account.vesting_ends_timestamp) - uint256(account.cliff_timestamp));
+            vested = x * y / z;
+        }
+        // the answer is how much is vested in total minute how much already withdrawn
+        return vested - uint256(account.amount_already_withdrawn_div1e12) * 1e12;
+    }
+
+    /**
+     * @notice User can claim their vested tokens.
+     */
+    function claim() external {
+        uint256 claim_amount = availableToClaim(msg.sender);
+        require(claim_amount > 0, "Can't claim 0 tokens");
+        accounts[msg.sender].amount_already_withdrawn_div1e12 = accounts[msg.sender].amount_already_withdrawn_div1e12 + uint64(claim_amount / 1e12);
+        _transferVotingUnits(msg.sender, address(0), (claim_amount / 1e12) * 1e12);
+        LOCKED_TOKEN.safeTransfer(msg.sender, (claim_amount / 1e12) * 1e12);
+    }
+
+    /**
+     * @notice User can donate tokens under vesting to DAO or other admin contract as us treasury.
+     */
+    function donateTokens(address toAdmin) external {
+        require(hasRole(DEFAULT_ADMIN_ROLE, toAdmin) == true, "invalid admin address");
+        uint256 balance = totalDeposited(msg.sender) - totalClaimed(msg.sender);
+        require(balance > 0, "nothing to donate");
+        accounts[msg.sender].amount_already_withdrawn_div1e12 = accounts[msg.sender].amount_already_withdrawn_div1e12 + uint64(balance / 1e12);
+        _transferVotingUnits(msg.sender, toAdmin, (balance / 1e12) * 1e12);
+        LOCKED_TOKEN.safeTransfer(toAdmin, (balance / 1e12) * 1e12);
+    }
+
+    /**
+    * @notice The owner of the contact can take away tokens accidentally sent to the contract.
+    */
+    function rescue(ERC20 token) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(token != LOCKED_TOKEN, "No allowed to rescue this token");
+        // allow to rescue ether
+        if (address(token) == address(0)) {
+            payable(msg.sender).transfer(address(this).balance);
+        } else {
+            token.safeTransfer(address(msg.sender), token.balanceOf(address(this)));
+        }
+    }
+
+    /**
+     * @dev Returns the amount under vesting for `account`.
+     */
+    function _getVotingUnits(address account) internal virtual override returns (uint256) {
```

Comment:
this is up to you, but I would generally try to change VestingTokenV2.sol only and add the all delegation voting logic there. So only one contact gets new logic, and we simply reuse existing battled-tested contact that holds $0.5M

Reply:
The  `Votes` contract must be inherited by the contract that performs all the balance manipulations.
The  `TimelockVaultVotes` contract performs all the balance manipulations.
This function needs to be in the contract that inherits the `Votes` contract.

Thus, we HAVE TO make changes to the TimelockVaultVotes contract. The other (bad) choice is to do code duplications, and override all balance manipulating functions in the VestingTokenV2.sol.